inputs/loadgenerator/core.go

```diff
@@ -8,58 +8,253 @@ import (
 	"context"
 	"fmt"
 	"os"
+	"sync"
 	"time"
 
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-libs/config"
 	"github.com/elastic/elastic-agent-libs/logp"
+
+	"go.uber.org/atomic"
+	"google.golang.org/protobuf/types/known/structpb"
 )
 
 type loadGenerator struct {
-	cfg    Config
-	now    func() time.Time
-	logger *logp.Logger
-}
-
-func (l loadGenerator) Run(ctx context.Context) error {
-	t := l.now()
-	delta := l.cfg.TimeDelta
-	delay := l.cfg.Delay
-
-	// Loop forever
-	if l.cfg.Loop {
-		l.logger.Debug("loop enabled")
-		for {
-			select {
-			case <-ctx.Done():
-				l.logger.Info("Shutdown signal received, closing loadGenerator")
-				return nil
-			default:
-				if err := l.send(l.next(t)); err != nil {
-					return fmt.Errorf("error sending event: %w", err)
+	now     func() time.Time
+	logger  *logp.Logger
+	runners map[string]*loadRunner
+	runMut  *sync.Mutex
+	events  atomic.Uint64
+}
+
+func newLoadGenerator(logger *logp.Logger) *loadGenerator {
+
+	return &loadGenerator{
+		now:     time.Now,
+		logger:  logger,
+		runners: make(map[string]*loadRunner),
+		runMut:  &sync.Mutex{},
+	}
+}
+
+// load runner handles the communications between a loadGenerator instance and the server
+type loadRunner struct {
+	stopFuncs map[string]chan struct{}
+	mapMut    *sync.Mutex
+	unit      *client.Unit
+}
+
+func newLoadRunner() *loadRunner {
+	return &loadRunner{
+		stopFuncs: make(map[string]chan struct{}),
+		mapMut:    &sync.Mutex{},
+	}
+}
+
+func (l *loadGenerator) Close(ctx context.Context) error {
+	// nothing here, for now
+	return nil
+}
+
+// Start the generator and initialize the client
+func (l *loadGenerator) Start(ctx context.Context) error {
+	agentClient, _, err := client.NewV2FromReader(os.Stdin, client.VersionInfo{
```

Comment:
Yah, unfortunately this is how all the inputs work under V2. elastic-agent sends off the config via stdin and it goes from there. I've been using the v2tool to deal with this, although the other alternative is just to bypass the V2 client init altogether.

Reply:
Yeah, I ended up "manually" configuring the client when I rebased my PR onto yours because the v2tool was not ready yet. And now I added a flag to choose between the config file and V2 protocol.